### PR TITLE
Horizons Designation

### PR DESCRIPTION
**Pull request by swalker2m**
Brings the `HorizonsDesignation` over from the old `ocs` project so that it is is available to use in the non-sidereal target model.  Horizons designation is a unique horizons id that can be used to construct non-sidereal target queries in JPL horizons.  This is mostly boilerplate but there were some small decisions that may need to be re-evaluated:

1) I placed the `HorizonsDesignation` in a new package called `target` anticipating that the rest of the target model will go there rather than `gem.math`.  Of course `target` is excluded by `.gitignore` by default so that may cause problems?

2) I made an `HorizonsType` enum in the database, anticipating that we'll want to store that, along with the string designation, in the non-sidereal target table.  It isn't really needed yet though.

### CODE REVIEW

File: modules/core/shared/src/main/scala/gem/enum/HorizonsType.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+sealed abstract class HorizonsType(
+  val tag: String,
+  val name: String
+)
+
+object HorizonsType {
+
+  /** @group Constructors */ case object Comet extends HorizonsType("Comet", "Comet")
+  /** @group Constructors */ case object AsteroidNew extends HorizonsType("AsteroidNew", "Asteroid New")
+  /** @group Constructors */ case object AsteroidOld extends HorizonsType("AsteroidOld", "Asteroid Old")
```
Comment (tpolecat) [inline modules/core/shared/src/main/scala/gem/enum/HorizonsType.scala]: These will eventually be shown to the user so `Asteroid (New Format)` or similar might be more helpful.

File: modules/core/shared/src/main/scala/gem/parser/HorizonsDesignation.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import gem.target.HorizonsDesignation._
+
+import atto._, Atto._
+import cats.implicits._
+
+/** Parser for [[gem.target.HorizonsDesignation]]. */
+trait HorizonsDesignationParsers {
+
+  private def des[A,B](s: String, p: Parser[A]): Parser[A] =
+    string(s"${s}_") ~> p
```
Comment (tpolecat) [inline modules/core/shared/src/main/scala/gem/parser/HorizonsDesignation.scala]: Probably don't need type param `B` here.

File: modules/core/shared/src/main/scala/gem/parser/HorizonsDesignation.scala [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+  private def textDes[A](s: String)(f: String => A): Parser[A] =
+    des(s, takeText.map(f))
+
+  private def numDes[A](s: String)(f: Int => A): Parser[A] =
+    des(s, int.map(f))
+
+  val comet: Parser[Comet] =
+    textDes("Comet"      )(Comet.apply      ) namedOpaque "comet"
```
Comment (tpolecat) [inline modules/core/shared/src/main/scala/gem/parser/HorizonsDesignation.scala]: whitespace

File: modules/core/shared/src/main/scala/gem/parser/HorizonsDesignation.scala [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+  val asteroidNew: Parser[AsteroidNew] =
+    textDes("AsteroidNew")(AsteroidNew.apply) namedOpaque "asteroidNew"
+
+  val asteroidOld: Parser[AsteroidOld] =
+    numDes ("AsteroidOld")(AsteroidOld.apply) namedOpaque "asteroidOld"
+
+  val major: Parser[MajorBody] =
```
Comment (tpolecat) [inline modules/core/shared/src/main/scala/gem/parser/HorizonsDesignation.scala]: Can we call this `majorBody` for consistency with the name?

File: modules/core/shared/src/main/scala/gem/parser/HorizonsDesignation.scala [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    numDes ("MajorBody"  )(MajorBody.apply  ) namedOpaque "majorBody"
+
+  val horizonsDesignation: Parser[HorizonsDesignation] =
+      (comet      .widen[HorizonsDesignation] |
+       asteroidNew.widen[HorizonsDesignation] |
+       asteroidOld.widen[HorizonsDesignation] |
+       major      .widen[HorizonsDesignation] ) named "horizonsDesignation"
```
Comment (tpolecat) [inline modules/core/shared/src/main/scala/gem/parser/HorizonsDesignation.scala]: This is irritating, sorry.

File: modules/core/shared/src/main/scala/gem/parser/HorizonsDesignation.scala (hunk 2 above)
Comment (swalker2m) [reply modules/core/shared/src/main/scala/gem/parser/HorizonsDesignation.scala]: Oops, that's left over from an earlier version.

**Comment by tpolecat** [comment]
The database enum makes sense to me because it guarantees that the designations will at least be well-formed, so 👍 on that.

**Comment by tpolecat** [comment]
I think the `target` package will end up being problematic, so we might reconsider that one. I have no good suggestion though.

**Comment by swalker2m** [comment]
> I had a few minor comments. Can we add a test that verifies round-tripping through the parser?

D'oh, that was ignored by `.gitignore`.  I'll force it in.

**Comment by tpolecat** [comment]
Yeah let's definitely rename `target/`

**Comment by tpolecat** [comment]
Honestly I would be fine with this in `gem`. We can subdivide when it gets too crowded. I was thinking of pushing `Enumerated` and `Location` into `gem.util` since they don't really have anything specific to do with the model.

**Comment by tpolecat** [comment]
Looks great! 👍

**Pull request merged**